Managers/EnemyManager.gd

```diff
@@ -2,7 +2,131 @@ extends Node
+	[3, 40, 60]: [],
+	[3, 60, 80]: [],
+	[3, 80, 100]: [],
+}
+
+## The group of enemies currently being used, set at the start of the combat [br]
+var current_enemy_group: EnemyGroup
+
+## The packed scene corresponding to the group of enemies, this is used to keep track of it between saves [br]
+## DO NOT FORGET to set this back to null once the fight is done [br]
+## The reset is handled via signals to be sure to not forget it when changing the battler
+var current_enemy_group_packed: PackedScene = null
+
+## The last recorded lower bound of the current sub-section
+var current_lower_bound: int = enemy_group_distribution.keys()[0][1]
+## The last recorded higher bound of the current sub-section
+var current_higher_bound: int = enemy_group_distribution.keys()[0][2]
+## The last used array of enemy group
+var enemy_group_array: Array[PackedScene]
+## The shuffled version of the `enemy_group_array`
+var shuffled_enemy_group_array: Array[PackedScene] = []
+
+func _ready() -> void:
+	if enemy_group_array.is_empty():
+		enemy_group_array.assign(enemy_group_distribution.values()[0])
+	PhaseManager.on_event_win.connect(_clear_data)
+	PhaseManager.on_defeat.connect(_clear_data)
+
+
+## This calls the inner function that actually chooses the enemy group [br]
+## This is done to save the packed scene in this EnemyManger before it is returned
+func choose_enemy_group() -> PackedScene:
+	if current_enemy_group_packed != null:
+		# if we load from a save
+		return current_enemy_group_packed
+	var packed_enemy_group: PackedScene = _choose_enemy_group_inner()
+	current_enemy_group_packed = packed_enemy_group
+	return packed_enemy_group
+
+## Chooses an enemy group in the following way: [br]
+## - Check if the player is in the same sub-section as the previous time this function was called [br]
+## - Choose an array of enemy group depending on player position if either sub-section changed, or the array to choose from is empty [br]
+## - Shuffle it, pop the last element of that array
+func _choose_enemy_group_inner() -> PackedScene:
+	if current_enemy_group != null:
+		return current_enemy_group_packed
+	if DebugVar.DEBUG_USE_TEST_ENEMY_GROUP:
+		return load("res://Entity/Enemy/EnemyGroup/test_group.tscn") # return a test group
+	var height_percent_position: float = MapManager.get_map_percent_with_player_position()
+	if (current_lower_bound <= height_percent_position && current_higher_bound > height_percent_position):
+		# meaning we stayed in the same sub-section of the map, keep the same array of enemies
+		if shuffled_enemy_group_array.is_empty():
+			shuffled_enemy_group_array = enemy_group_array.duplicate()
+			shuffled_enemy_group_array.shuffle()
+
+		return shuffled_enemy_group_array.pop_back()
+	else:
+		# we entered a new sub-section, search the corresponding array of enemy group
+		for key: Array in enemy_group_distribution:
+			# include lower bound, exclude higher bound
+			if (key[1] <= height_percent_position && key[2] > height_percent_position):
```

Comment:
One thing came to mind but I am not sure if it should be part of this PR as we don't even have other floors implemented yet as far as I know, but currently this is only checking for the height bounds, not which floor we're on (key[0])
If we don't have floors implemented yet in any way then it probably can't really be added now - or can we have some placeholder at least so that it's not forgotten ?

Reply:
We check the section we are one, because we do `for key in enemy_group_distribution` and later `var new_enemy_group_array: Array = enemy_group_distribution[key]`
Since the key is of the shape `[floor, lower_bound, upper_bound]`, we check the section we are on

Reply:
I know that there is support for having floors in the distribution, just that we are not actually checking for the floors in any way within the function - although as they are not yet implemented best we could do would probably just be some sort of placeholder for now anyway so it shouldn't really matter